character-controller.js

```diff
@@ -690,34 +733,34 @@ class StatePlayer extends PlayerBase {
             if (!cancelFn.isLive()) return;
             _setNextAvatarApp(nextAvatarApp);
           } else {
-            console.warn('switching avatar to instanceId that does not exist in any app manager', instanceId);
+            _setNextAvatarApp(null)
           }
         }
       }
-    } else {
-      _setNextAvatarApp(null);
     }
-    
+
     this.syncAvatarCancelFn = null;
   }
   setSpawnPoint(position, quaternion) {
-    const localPlayer = metaversefile.useLocalPlayer();
-    localPlayer.position.copy(position);
-    localPlayer.quaternion.copy(quaternion);
+    this.position.copy(position);
+    this.quaternion.copy(quaternion);
 
     camera.position.copy(position);
     camera.quaternion.copy(quaternion);
     camera.updateMatrixWorld();
 
-    if (this.characterController) {
+    if (this.characterPhysics) {
       this.characterPhysics.setPosition(position);
     }
   }
   getActions() {
     return this.getActionsState();
   }
   getActionsState() {
-    let actionsArray = this.playerMap.has(avatarMapName) ? this.playerMap.get(actionsMapName, Z.Array) : null;
+    if (!this.playerMap) return []
+    let actionsArray = this.playerMap.has(avatarMapName)

```

Comment:
This should go through a network abstraction so that this code is not aware of Z.js.

Reply:
It could be a combination of `PlayerNetworkBinding` and `this.player.state = new PlayerState()` which can handle the implementation of things like actions.

Reply:
The point is to make `Player` less prone to race conditions since it is dealing in very low level asynchronous manipulation.